packages/kit/src/runtime/client/client.js

```diff
@@ -0,0 +1,1233 @@
+	let current = {
+		// @ts-ignore - we need the initial value to be null
+		url: null,
+		session_id: 0,
+		branch: []
+	};
+
+	let started = false;
+	let autoscroll = true;
+	let updating = false;
+	let session_id = 1;
+
+	/** @type {Promise<void> | null} */
+	let invalidating = null;
+
+	/** @type {import('svelte').SvelteComponent} */
+	let root;
+
+	/** @type {App.Session} */
+	let $session;
+
+	let ready = false;
+	stores.session.subscribe(async (value) => {
+		$session = value;
+
+		if (!ready) return;
+		session_id += 1;
+
+		const intent = get_navigation_intent(new URL(location.href));
+		update(intent, [], true);
+	});
+	ready = true;
+
+	/** Keeps tracks of multiple navigations caused by redirects during rendering */
+	let navigating = 0;
+
+	let router_enabled = true;
+
+	// keeping track of the history index in order to prevent popstate navigation events if needed
+	let current_history_index = history.state?.[INDEX_KEY] ?? 0;
+
+	if (current_history_index === 0) {
+		// create initial history entry, so we can return here
+		history.replaceState({ ...history.state, [INDEX_KEY]: 0 }, '', location.href);
+	}
+
+	// if we reload the page, or Cmd-Shift-T back to it,
+	// recover scroll position
+	const scroll = scroll_positions[current_history_index];
+	if (scroll) scrollTo(scroll.x, scroll.y);
+
+	let hash_navigating = false;
+
+	/** @type {import('types').Page} */
+	let page;
+
+	/** @type {{}} */
+	let token;
+
+	/** @type {{}} */
+	let navigating_token;
+
+	/**
+	 * @param {string} href
+	 * @param {{ noscroll?: boolean; replaceState?: boolean; keepfocus?: boolean; state?: any }} opts
+	 * @param {string[]} redirect_chain
+	 */
+	async function goto(
+		href,
+		{ noscroll = false, replaceState = false, keepfocus = false, state = {} },
+		redirect_chain
+	) {
+		const url = new URL(href, get_base_uri(document));
+
+		if (router_enabled) {
+			return navigate({
+				url,
+				scroll: noscroll ? scroll_state() : null,
+				keepfocus,
+				redirect_chain,
+				details: {
+					state,
+					replaceState
+				},
+				accepted: () => {},
+				blocked: () => {}
+			});
+		}
+
+		await traditional_navigation(url);
+	}
+
+	/** @param {URL} url */
+	async function prefetch(url) {
+		if (!owns(url)) {
+			throw new Error('Attempted to prefetch a URL that does not belong to this app');
+		}
+
+		const intent = get_navigation_intent(url);
+
+		loading.promise = get_navigation_result(intent, false);
+		loading.id = intent.id;
+
+		return loading.promise;
+	}
+
+	/**
+	 * @param {import('./types').NavigationIntent} intent
+	 * @param {string[]} redirect_chain
+	 * @param {boolean} no_cache
+	 * @param {{hash?: string, scroll: { x: number, y: number } | null, keepfocus: boolean}} [opts]
+	 */
+	async function update(intent, redirect_chain, no_cache, opts) {
+		const current_token = (token = {});
+		let navigation_result = await get_navigation_result(intent, no_cache);
+
+		if (!navigation_result && intent.url.pathname === location.pathname) {
```

Comment:
something that might be worth a comment - I'm not understanding the second half of this check. if you want to navigate to the page we're already on, isn't that essentially just a no-op? why do we want to show a 404 page in that case?

Reply:
added a comment